### PR TITLE
fix(runtime): carve the Cayenne compaction memory pool only when a dataset can compact into it (fixes #12320)

### LINKED ISSUE #12320: Cayenne: only reserve compaction and mem-tier memory when eligible datasets exist
## Problem

The Cayenne compaction pool and in-memory CDC tier budgets are installed whenever *any* enabled Cayenne dataset exists, with no check that either subsystem can actually run. The gate is `cayenne_configured` (`crates/runtime/src/builder.rs:476-494`), which reads only `accel.enabled` and `accel.engine == "cayenne"`.

Neither budget allocates RSS — both are counters. But the compaction carve is subtracted straight out of the query memory limit (`crates/runtime/src/datafusion/builder.rs:756`), so it directly reduces what queries may reserve.

On a full-refresh deployment (8 Cayenne file-mode datasets, 40 GiB pod, `memory_limit: 32GiB`), profiled with heaptrack across 9 runs:

- query pool 25.6 GiB — **6.4 GiB carved for compaction**
- mem-tier budget 4.0 GiB — **`used = 0` for the entire process lifetime**
- compaction absent from the top-150 peak allocators

The mem-tier is not merely idle, it is **unreachable**: `CdcDurability::Memory` is eligibility-gated (`crates/runtime/src/dataaccelerator/cayenne/mod.rs:1304`) and downgraded to `File` for any dataset off the small-write refresh profile, so the `try_reserve_bytes` call site is never entered.

No configuration reclaims either budget:

| lever | result |
|---|---|
| `cayenne_cdc_durability: file` | no-op — already forced; budget still installed |
| `cayenne_compaction_background_interval_ms: 0` | disables the compactor but **keeps the carve** — the gate never reads the param |
| `cayenne_compaction_memory_fraction: 0.05` | floor only: 6.4 → 1.6 GiB |
| `dedicated_thread_pool: disabled` | removes both, but also disables the dedicated SQL engine thread pool |

Lowering the compaction fraction does not reduce the total either — the mem-tier is the *remainder* and absorbs whatever compaction releases.

## Proposed change

Install each budget only when at least one enabled Cayenne dataset is eligible, reusing the predicate that already decides the mem-tier downgrade (`uses_small_write_refresh_profile`, `dataaccelerator/cayenne/mod.rs:718`): `refresh_mode` `changes`/`caching`, or `append`

1. **Compaction memory reservation** — gate `compaction_memory_fraction` (`builder.rs:494`) on eligible datasets, so no carve and no dedicated compaction runtime when nothing can compact.
2. **Mem-tier reservation** — gate `mem_tier_budget_bytes` (`datafusion/builder.rs:774`) the same way. `datafusion/mod.rs:1626` is already `if let Some(..)`, so `None` cleanly skips the budget, the query-pool mirror, and the dynamic re-partition sampler.

### Tracing must follow the gate

Startup logs currently announce the tier budget (`datafusion/mod.rs:1636`) and the compaction runtime (`builder.rs:574`) as "active". With these gates those paths are skipped, so:

- **Skipped** → one INFO per subsystem naming the reason, e.g. *"Cayenne compaction memory pool not reserved: no dataset uses a small-write refresh profile (changes/caching, or append with refresh_check_interval <= 5m); N Cayenne dataset(s) configured."*
- **Installed** → keep the existing lines and state how many datasets made it eligible, so the number is auditable against the spicepod.
- The coordination line already prints `query_memory_pool_bytes`, `compaction_memory_bytes`, `mem_tier_budget_bytes`, `total_memory` — each must still match the installed values under every gate combination. Reporting a budget that was never installed is worse than logging nothing.
- Gauges `cayenne_mem_tier_budget_total_bytes` / `_used_bytes` (`dataaccelerator/cayenne/mod.rs:411,423`) must read 0 when skipped, so `total > 0, used == 0` stays a reliable "reserved but idle" signature.

### Considerations

Dynamically created tables might be affected - what if  new table created dynamically with new cayenne append/changes configuration. Out of scope scenario, can be improved later

### PR DESCRIPTION
**Pull request by claudespice**
Fixes #12320

## Summary

The Cayenne compaction memory carve was taken whenever *any* enabled Cayenne acceleration existed. The gate was `cayenne_configured`, which reads only `accel.enabled` and `accel.engine == "cayenne"` — it never checked that anything could compact into the carve.

The carve is a counter, not an allocation, but it is subtracted straight out of the query memory limit, so it directly reduces what queries may reserve. On the deployment profiled in #12320 — 8 Cayenne file-mode datasets on full refresh, 40 GiB pod, `memory_limit: 32GiB` — that meant **6.4 GiB carved for a compaction pool nothing compacts into**, with compaction absent from the top-150 peak allocators. No parameter reclaimed it: `cayenne_compaction_background_interval_ms: 0` disables the compactor but keeps the carve, `cayenne_compaction_memory_fraction` is a floor, and only `dedicated_thread_pool: disabled` released it — at the cost of the dedicated SQL engine thread pool.

## Changes

Take the carve only when at least one enabled Cayenne acceleration can compact into it: **a file acceleration mode AND the small-write refresh profile** (`refresh_mode` `changes`/`caching`, or `append` with `refresh_check_interval <= 5m`).

Both halves matter, and `mode` is the one that is easy to miss:

- **File mode.** `mode: memory` is the Spicepod *default*, and `apply_memory_mode_overrides` sets `compaction_background_interval_ms = 0` — such a table never compacts at all. Gating on the refresh profile alone would still have carved for a default-mode `refresh_mode: changes` dataset, the most common CDC shape.
- **Small-write refresh profile.** The profile the carve was sized for: continuous small writes producing the small files tiered compaction exists to merge. Reuses the predicate that already decides the `cdc_durability` downgrade, refactored so both callers share one definition rather than restating the rule.

Also:

- Log which way the gate went — one INFO naming the reason when skipped, and the eligible count when taken, so the number is auditable against the spicepod.
- Move `cayenne_active` to its own `DataFusionBuilder` field instead of deriving it from `compaction_memory_fraction.is_some()`. It also drives the reduced Cayenne query-memory default, the spill-directory hint, and the in-memory CDC tier budget; without this split, gating the carve would have silently changed all three.

Views are counted alongside datasets. None can qualify today (`ViewBuilder::try_from` rejects every view refresh mode except `full`), but #12189 proposes `refresh_mode: append` for views, and a datasets-only count would quietly become wrong when it lands.

## What is deliberately NOT gated

**The in-memory CDC tier budget stays on `cayenne_active`.** #12320 asked for it to be gated the same way; it should not be, because the two budgets fail in opposite directions:

| | costs the query pool | omitting it |
|---|---|---|
| compaction carve | yes — subtracted from the memory limit | **fail-safe**: compaction accounts against the query pool |
| in-memory CDC tier budget | no — sized from what the query pool *leaves over* (`coordinated_mem_tier_budget`) | **fail-open**: `try_reserve_bytes` succeeds unconditionally with no budget installed |

So gating the tier budget reclaims nothing for queries and drops the aggregate cap that keeps memory-mode CDC within host RAM, leaving only per-table caps. All of the profiled 6.4 GiB comes from the carve. Keeping the tier on `cayenne_active` also means a spicepod hot reload that adds a CDC dataset still finds the cap installed (originally filed as #12325, closed with this reasoning).

## Behavior notes

**Compaction still runs.** Its background interval is non-zero by default for every file-mode Cayenne table, so this does not disable compaction — `COMPACTION_RUNTIME_ENV` falls back to the shared query environment when unset, exactly as it does when Cayenne is absent or dedicated thread pools are disabled. The deliberate trade for a file-mode bulk-refresh pod: it loses compaction's *isolation* from the query pool and gains the carved bytes back. That is what #12320 profiled and asked for.

**The dedicated compaction *thread pool* is unchanged.** #12320 also asked for no compaction runtime when nothing can compact, but that runtime is not driven by this gate: `bin/spiced` creates it whenever dedicated thread pools are enabled, deliberately, so Cayenne can be activated lazily after startup. Left alone.

## Test plan

- `make lint`
- `make test`

New unit tests in `crates/runtime/src/builder.rs`:

- `count_cayenne_budget_eligible_accelerations_counts_only_eligible_accelerations` — every file mode (`file`/`file_create`/`file_update`) counts; `mode: memory` never does on any refresh mode; the three eligible refresh shapes count, including `append` at exactly the 5m threshold; `append` past the threshold, with no interval, or with an unparseable interval does not; bulk refresh modes and the `full` default do not; disabled accelerations and other engines do not; engine matching stays case-insensitive; eligible accelerations accumulate and an ineligible sibling does not mask an eligible one; views are traversed; the absent-app case is 0.
- `cayenne_memory_budgets_are_reserved_only_for_an_eligible_dataset` — the #12320 regression: a full-refresh pod and a default-`mode: memory` CDC pod are both `cayenne_active` with **no** carve; one eligible dataset among ineligible siblings restores it; an explicit `cayenne_compaction_memory_fraction` is still honored and clamped; and the pre-existing gates are unchanged (`dedicated_thread_pool: disabled` drops both the carve and `cayenne_active`, a non-Cayenne pod was never active).

Neuter-verified twice, both against the regression test: dropping the eligibility term reports `Some(0.2)` where `None` is required (trunk's 20% carve — the reported bug), and dropping the file-mode term reports the `mode: memory` case as eligible (`left: 1, right: 0`).

## Checklist

- [x] Fix is minimal and scoped to the reported defect
- [x] Regression test fails on the old gate, passes on the new one
- [x] Edge cases covered (every acceleration mode, every refresh mode, threshold boundary, unparseable interval, disabled/other-engine, views, absent app)
- [x] No behavior change for eligible deployments, for `dedicated_thread_pool: disabled`, for the Cayenne query-memory default, or for the in-memory CDC tier budget
- [x] Windows build preserved (`dataaccelerator::cayenne` is `cfg(not(windows))`; the counter is cfg-split)
- [x] No `.unwrap()`; `tracing::` macros only; log messages single-line
- [x] `docs/cayenne/cayenne.md` reviewed — the change is in `crates/runtime` and the doc covers neither budget's install conditions, so no update needed

### CODE REVIEW

**Comment by claudespice** [comment]
@copilot review

File: crates/runtime/src/dataaccelerator/cayenne/mod.rs [hunk 1]
```diff
@@ -715,11 +715,29 @@ fn apply_refresh_mode_defaults(
     }
 }
 
+/// Whether the dataset's refresh mode writes small batches continuously, which is
+/// what the CDC-shaped defaults (inline memtable, aggressive compaction triggers,
+/// `cdc_durability: memory`) are for. Also decides whether the process installs the
+/// Cayenne compaction memory carve and the in-memory CDC tier budget at all — see
+/// `builder::count_cayenne_budget_eligible_accelerations`.
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime/src/dataaccelerator/cayenne/mod.rs:722]: Agreed, and the wording was backwards in a way this PR is specifically about. `uses_small_write_refresh_profile` gates only the compaction carve; the in-memory CDC tier budget is gated on `cayenne_active` alone (`datafusion/builder.rs`: `mem_tier_budget_bytes = self.cayenne_active.then(...)`), so it is installed for every Cayenne deployment whatever its refresh mode.

Fixed in dab31c2, and the doc now also gives the reason the two differ: the carve takes bytes away from queries, so it is only worth taking when something can compact into it, whereas the tier budget is sized from what the query pool leaves over and costs queries nothing.

File: crates/runtime/src/builder.rs [hunk 2]
```diff
@@ -1094,22 +1083,156 @@ fn parse_usize_runtime_param(params: &HashMap<String, String>, key: &str) -> Opt
...
+    // reserved budgets against the spicepod. Silent for a non-Cayenne deployment.
+    if cayenne_active {
+        if compaction_memory_fraction.is_some() {
+            tracing::info!(
+                eligible_accelerations,
+                "Reserving the Cayenne compaction memory pool: {eligible_accelerations} dataset(s) can compact into it."
+            );
+        } else {
+            tracing::info!(
+                "Cayenne compaction memory pool not reserved: no dataset can compact into it (needs a file acceleration mode AND refresh_mode changes/caching, or append with refresh_check_interval <= 5m). The full query memory limit stays available to queries, and compaction accounts against the query pool."
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime/src/builder.rs]: Both points are right.

On the memory limit: with `cayenne_active` true and `runtime.query.memory_limit` unset, `effective_query_memory_limit` derives the default from `CAYENNE_QUERY_MEMORY_PERCENT` of host rather than leaving the pool whole, so "the full query memory limit stays available to queries" only holds for an explicit limit. The line now says nothing is carved off for compaction (which accounts against the query pool instead) and that the pool itself still follows the Cayenne default split unless the limit is set explicitly.

On "dataset": `count_cayenne_budget_eligible_accelerations` walks `spicepod_accelerations`, which chains `app.datasets` and `app.views`, so both messages now say "acceleration(s)". Fixed in dab31c2.

File: crates/runtime/src/datafusion/builder.rs [hunk 3]
```diff
@@ -642,15 +655,20 @@ impl DataFusionBuilder {
         self
     }
 
-    /// Carve a dedicated compaction memory pool of `fraction` of the query
-    /// memory limit. Set by the Runtime builder only when Cayenne acceleration
-    /// is configured and dedicated thread pools are enabled.
+    /// See [`Self::compaction_memory_fraction`].
     #[must_use]
     pub fn compaction_memory_fraction(mut self, fraction: Option<f64>) -> Self {
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime/src/datafusion/builder.rs]: Correct — `Self::compaction_memory_fraction` resolves to the setter it is written on, not to the field it meant to point at, and the same applies to `Self::cayenne_active` on the next setter. Both now describe what the setter does and who sets it (the Runtime builder), including the point that `cayenne_active` does not depend on any dataset being compaction-eligible. Fixed in dab31c2.

**Review by lukekim — APPROVED** [review]
